java/src/jni/h5util.c

```diff
@@ -4078,335 +4076,297 @@ translate_rbuf(JNIEnv *env, jobjectArray ret_buf, jlong mem_type_id, H5T_class_t
     jmethodID doubleValueMid =
         ENVPTR->GetStaticMethodID(ENVONLY, cDouble, "valueOf", "(D)Ljava/lang/Double;");
 
-    if (type_class == H5T_VLEN) {
-        if (!(memb = H5Tget_super(mem_type_id)))
-            H5_LIBRARY_ERROR(ENVONLY);
-        if ((vlClass = H5Tget_class(memb)) < 0)
-            H5_LIBRARY_ERROR(ENVONLY);
-        if (!(vlSize = H5Tget_size(memb)))
-            H5_LIBRARY_ERROR(ENVONLY);
+    if (!(typeSize = H5Tget_size(mem_type_id)))
+        H5_LIBRARY_ERROR(ENVONLY);
 
-        /* Convert each element to a list */
-        for (i = 0; i < (size_t)count; i++) {
-            hvl_t    vl_elem;
-            jboolean found_jList = JNI_TRUE;
-
-            /* Get the number of sequence elements */
-            HDmemcpy(&vl_elem, (char *)raw_buf + i * sizeof(hvl_t), sizeof(hvl_t));
-            jsize nelmts = (jsize)vl_elem.len;
-            if (vl_elem.len != (size_t)nelmts)
-                H5_JNI_FATAL_ERROR(ENVONLY, "translate_rbuf: overflow of number of VL elements");
-
-            if (nelmts < 0)
-                H5_BAD_ARGUMENT_ERROR(ENVONLY, "translate_rbuf: number of VL elements < 0");
-
-            /* The list we're going to return: */
-            if (NULL == (jList = ENVPTR->GetObjectArrayElement(ENVONLY, (jobjectArray)ret_buf, (jsize)i))) {
-                found_jList = JNI_FALSE;
-                if (NULL == (jList = (jobjectArray)ENVPTR->NewObject(ENVONLY, arrCList, arrListMethod, 0)))
-                    H5_OUT_OF_MEMORY_ERROR(ENVONLY, "translate_rbuf: failed to allocate list read buffer");
-            }
+    switch (type_class) {
+        case H5T_VLEN: {
+            if (!(memb = H5Tget_super(mem_type_id)))
+                H5_LIBRARY_ERROR(ENVONLY);
+            if ((vlClass = H5Tget_class(memb)) < 0)
+                H5_LIBRARY_ERROR(ENVONLY);
+            if (!(vlSize = H5Tget_size(memb)))
+                H5_LIBRARY_ERROR(ENVONLY);
+
+            /* Convert each element to a list */
+            for (i = 0; i < (size_t)count; i++) {
+                hvl_t        vl_elem;
+                jboolean     found_jList = JNI_TRUE;
+                int          ret_buflen  = -1;
+                jobjectArray jList       = NULL;
+
+                /* Get the number of sequence elements */
+                HDmemcpy(&vl_elem, (char *)raw_buf + i * sizeof(hvl_t), sizeof(hvl_t));
+                jsize nelmts = (jsize)vl_elem.len;
+                if (vl_elem.len != (size_t)nelmts)
+                    H5_JNI_FATAL_ERROR(ENVONLY, "translate_rbuf: overflow of number of VL elements");
+
+                if (nelmts < 0)
+                    H5_BAD_ARGUMENT_ERROR(ENVONLY, "translate_rbuf: number of VL elements < 0");
+
+                ret_buflen = ENVPTR->GetArrayLength(ENVONLY, ret_buf);
```

Comment:
No need to do this multiple times for each loop.

Reply:
Right - I missed that.

Reply:
But then again if ret_buf is only partially inited, I should get the existing members before we create new ones?

Reply:
Will move the GetArrayLength call out of the loop.